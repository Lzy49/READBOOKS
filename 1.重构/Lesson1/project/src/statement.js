import { createStatementData } from "./createStatementData.js";
// create text statement
export function statement(invoice, plays) {
  const statementData = createStatementData(invoice, plays);
  return renderPlainText(statementData);
}
function renderPlainText(data) {
  let result = `Statement for ${data.customer}\n`;
  for (let perf of data.performances) {
    result += ` ${perf.play.name}:${usd(perf.amount)}(${
      perf.audience
    } seats)\n`;
  }
  result += `Amount owed is ${usd(data.totalAmount)}\n`;
  result += `You earned ${data.totalVolumeCredits} credits\n`;
  console.log(result);
  return result;
}
// create html statement
export function htmlStatement(invoice, plays) {
  return renderPlainHTML(createStatementData(invoice, plays));
}
function renderPlainHTML(data) {
  let result = `<h1>Statement for ${data.customer}</h1>\n`;
  result += `<table>\n`;
  result += `<tr><th>player</th><th>amount</th><th>audience</th></tr>`;
  for (let perf of data.performances) {
    result += `<tr><td>${perf.play.name}</td><td>${usd(perf.amount)}</td><td>${
      perf.audience
    } seats</td></tr></td>\n`;
  }
  result += `</table> \n`;
  result += `<p>Amount owed is ${usd(data.totalAmount)}</p>\n`;
  result += `<p>You earned ${data.totalVolumeCredits} credits</p>\n`;
  console.log(result);
  return result;
  function usd(aNumber) {
    return new Intl.NumberFormat("en-US", {
      style: "currency",
      currency: "USD",
      minimumFractionDigits: 2,
    }).format(aNumber / 100);
  }
}
function usd(aNumber) {
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: 2,
  }).format(aNumber / 100);
}
