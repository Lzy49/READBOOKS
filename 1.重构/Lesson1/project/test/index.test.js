import {describe,it,expect} from 'vitest'
import invoice from "./invoices.js";
import play from "./play.js";
import statement from "./main.js";
describe("happy path", () => {
  it("test",()=>{
    // 第一步创建可靠的测试
    expect(statement(invoice, play)).toMatchSnapshot()
  })
});
