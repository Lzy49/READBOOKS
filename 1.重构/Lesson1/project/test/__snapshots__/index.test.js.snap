// Vitest Snapshot v1

exports[`happy path > test 1`] = `
"Statement for BigCo
 Hamlet:$650.00(55 seats)
 As You like It:$490.00(35 seats)
 Othello:$500.00(40 seats)
Amount owed is $1,640.00
You earned 47 credits
"
`;
